### ISSUE TITLE
Acme::Client::Error::Unauthorized: No registration exists matching provided key

### ISSUE THREAD

**Issue by hdiwan** [lgromanowski/letsencrypt-plugin#12]
I've generated a key myself and can't figure out how to tie it to letsencrypt. Help?

**Comment by lgromanowski** [comment]
Hi,
if you generated your key in similar way like:
```
$ openssl genrsa 4096 keyfile.pem
```
then please create a new directory called ie. ```key``` in your Rails application directory and than please  copy ```keyfile.pem``` file into it. After that please modify ```private_key``` variable in ```config/letsencrypt_plugin.yml``` configuration file accordingly.

**Comment by hdiwan** [comment]
That's what I did. And then I got the error. -- H

On 2 January 2016 at 01:21, Lukasz Gromanowski <notifications@github.com>
wrote:

> Hi,
> if you generated your key in similar way like:
>
> $ openssl genrsa 4096 keyfile.pem
>
> then please create a new directory called ie. key in your Rails
> application directory and than please copy keyfile.pem file into it.
> After that please modify private_key variable in
> config/letsencrypt_plugin.yml configuration file accordingly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lgromanowski/letsencrypt-plugin/issues/12#issuecomment-168375000>
> .
>



-- 
OpenPGP: https://hasan.d8u.us/gpg.key
Sent from my mobile device
Envoyé de mon portable

**Comment by lgromanowski** [comment]
Unfortunately I can't reproduce this issue. Could you please post ```config/letsencrypt_plugin.yml```, and output of ```ls -al``` inside keyfile directory?

**Comment by hdiwan** [comment]
letencrypt_plugin.yml:

default: &default

  endpoint: "https://acme-v01.api.letsencrypt.org/"

  email: "hasandiwan@gmail.com"

  domain: "example.com"

  private_key: "key/keyfile.pem"                            # in Rails.root

  output_cert_dir: "certificates"                           # in Rails.root


production:

  <<: *default


development:

  <<: *default


test:

  <<: *default


% ls -al .rails/ssl/key

total 16

drwxrwxr-x  2 ubuntu ubuntu 4096 Jan  2 00:02 .

drwxrwxr-x 12 ubuntu ubuntu 4096 Jan  2 00:02 ..

-rw-rw-r--  1 ubuntu ubuntu 6363 Jan  2 00:02 keyfile.pem

On 4 January 2016 at 11:42, Lukasz Gromanowski <notifications@github.com>
wrote:

> Unfortunately I can't reproduce this issue. Could you please post
> config/letsencrypt_plugin.yml, and output of ls -al inside keyfile
> directory?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lgromanowski/letsencrypt-plugin/issues/12#issuecomment-168780118>
> .
>



-- 
OpenPGP: https://hasan.d8u.us/gpg.key
Sent from my mobile device
Envoyé de mon portable

**Comment by lgromanowski** [comment]
You have ```private_key``` set to ```key/keyfile.pem``` but you are showing the ```ls``` output from ```.rails/ssl/key```. Did you copy that key into ```you_rails_app/key``` directory? Could you please try to generate a new key inside ```you_rails_app/key``` directory and check if it is working?

**Comment by hdiwan** [comment]
my rails_app is in .rails/ssl and it's still not working.

On 4 January 2016 at 14:23, Lukasz Gromanowski <notifications@github.com>
wrote:

> You have private_key set to key/keyfile.pem but you are showing the ls
> output from .rails/ssl/key. Did you copy that key into you_rails_app/key
> directory? Could you please try to generate a new key inside
> you_rails_app/key directory and check if it is working?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lgromanowski/letsencrypt-plugin/issues/12#issuecomment-168829839>
> .
>



-- 
OpenPGP: https://hasan.d8u.us/gpg.key
Sent from my mobile device
Envoyé de mon portable

**Comment by lgromanowski** [comment]
Did you change domain name ```domain: "example.com"``` to some domain name which you own?

**Comment by hdiwan** [comment]
No, do I need to have generated the certificate using letsencrypt?

**Comment by lgromanowski** [comment]
As I mentioned in earlier comment I can't reproduce this problem. Could you please write here what are you doing step-by-step (commands + commands output) from beginning till you encounter this error (please use github code [markdown](https://guides.github.com/features/mastering-markdown/) for it)?

**Comment by hdiwan** [comment]
I'm following instructions, I checked the source out from github, created a gemset, hit bundle install. My letsencrypt_plugin.yml consists of the following:

`
default: 
  endpoint: "https://acme-v01.api.letsencrypt.org/"
  email: "hasandiwan@gmail.com"
  domain: "prolificprogrammer.com"
  private_key: "key/keyfile.pem"                            # in Rails.root
  output_cert_dir: "certificates"                           # in Rails.root

production:
  <<: *default

development:
  <<: *default

test:
  <<: *default`

I've noticed that you have no unit tests, so I haven't run those -- I'd like to write a few when I move beyond this problem, if that's ok. 

I start the server, open another terminal and hit rake letsencrypt_plugin and get that error. The problem could be a case of misunderstanding the point of the web application. Is its scope restricted to renewing the certificates every 90 days automatically or can I get the initial certificate from your application as well?

**Comment by lgromanowski** [comment]
You can get both - new certificates and you can renew using this script as well. 

You mentioned that you are cloned source from github and created gemset - could you please try to remove letsencrypt_plugin gem first by invoking: ```gem uninstall letsencrypt_plugin``` (remove all versions) and then add it to Gemfile, ie:
```ruby
source 'https://rubygems.org'

# ... some other gems ...

gem 'letsencrypt_plugin'
```
and use ```bundle install``` to get release version of letsencrypt_plugin instead of using git sources?

**Comment by hdiwan** [comment]
Still, the same error. I put my code at
https://bitbucket.org/hd1/letencrypt-online if you'd like to take a look.
-- H

On 6 January 2016 at 13:23, Lukasz Gromanowski <notifications@github.com>
wrote:

> You can get both - new certificates and you can renew using this script as
> well.
>
> You mentioned that you are cloned source from github and created gemset -
> could you please try to remove letsencrypt_plugin gem first by invoking: gem
> uninstall letsencrypt_plugin (remove all versions) and then add it to
> Gemfile, ie:
>
> source 'https://rubygems.org'
> # ... some other gems ...
> gem 'letsencrypt_plugin'
>
> and use bundle install to get release version of letsencrypt_plugin
> instead of using git sources?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lgromanowski/letsencrypt-plugin/issues/12#issuecomment-169466813>
> .
>



-- 
OpenPGP: https://hasan.d8u.us/gpg.key
Sent from my mobile device
Envoyé de mon portable

**Comment by lgromanowski** [comment]
Ok, it seems that the problem is in key size - acme server (letsencrypt.org) doesn't allow key size > 4096 (it shows quite strange error message). Please try with key size = 4096, it should work.

**Comment by hdiwan** [comment]
Now I'm getting "challenge verification failed"; I have to run now, will
update more later. -- H

On 6 January 2016 at 14:03, Lukasz Gromanowski <notifications@github.com>
wrote:

> Ok, it seems that the problem is in key size - acme server (
> letsencrypt.org) doesn't allow key size > 4096 (it shows quite strange
> error message). Please try with key size = 4096, it should work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lgromanowski/letsencrypt-plugin/issues/12#issuecomment-169476873>
> .
>



-- 
OpenPGP: https://hasan.d8u.us/gpg.key
Sent from my mobile device
Envoyé de mon portable

**Comment by backwardm** [comment]
Hello... first of all, thanks for making this gem. It's super useful and I love how you've implemented the engine to listen on top of a running app. Nice work!

Secondly, I ran into the issue that @hdiwan is describing here. (same error message) In my case, I cloned a production VPS and pointed an extra domain name to the clone for testing, but when specifying which domain to secure on the clone in the .yml file, I used the production domain name that the original server was still using. So letsencrypt was trying to validate via my production server instead of my cloned test server.

Simple mistake, but maybe that's what @hdiwan is seeing as well?
